src/main/scala/mesosphere/marathon/core/launcher/impl/InstanceOpFactoryImpl.scala

```diff
@@ -303,7 +302,8 @@ class InstanceOpFactoryImpl(
         val (executorInfo, groupInfo, hostPorts) = TaskGroupBuilder.build(pod, offer,
           instanceId, podContainerTaskIds, builderConfig, runSpecTaskProc, resourceMatch, Some(volumeMatch))
 
-        val stateOp = InstanceUpdateOperation.Provision(reservedInstance.provisioned(agentInfo, hostPorts, pod, podContainerTaskIds, clock.now()))
+        val now = clock.now()
+        val stateOp = InstanceUpdateOperation.Provision(reservedInstance.instanceId, agentInfo, pod.version, Tasks.provisioned(podContainerTaskIds, agentInfo, hostPorts, pod, now), now)
```

Comment:
Hm, this is kind of mixing two things here. IMHO `InstanceUpdateOperation.Provision` is either an update or and command. If we make it a command I'd sent `Provision(reservedInstance.instanceId, agentInfo, version, taskIds, hostPorts, now)` and call `Tasks.provisioned()` in the update.

That said, I'm afraid the issue is structural, ie they way we update. The question for me is why are we overriding? As I see it we should set the goal and provision in sequence per instance id. This would avoid all these races we keep hunting. So something like

```scala
class Scheduler  {

val queue = Source.queue[Commands]
  .groupBy(16, cmd => Math.abs(cmd.instanceId.idString.hashCode) % 16)
  .mapAsync(1) {
     case ProcessMesosOffer(offer) => 
        val match = OfferMatcher(...)
        instanceTracker.update(match)
     case UpdateInstance(goal) => instanceTracker.update(...)
   }
  .mergeSubstreams

}

def processOffer(offer) = queue.offer(ProcessMesosOffer(offer, ...))

def setGoal(goal) = queue.offer(UpdateInstance(goal, ...))
```

I know it would not be possible right now but it is kind of how I imagined the architecture. Maybe going closer to what @timcharper proposed.

Anyways, coming back to this diff. We change the semantics of `InstanceUpdateOperation.Provision`. I never really liked that pattern and wish it would be a simple update. So what is the real issue here? That we matched something that had its goal changed in the mean time? IMHO that is something the `TaskLauncherActor` or `OfferMatcherManagerActor` should deal with.